### ISSUE TITLE
Multi-GPU Memory-Usages are not equal. It is higher for the first one which becomes the bottleneck.

### ISSUE THREAD

**Issue by pocketvision** [ultralytics/yolov5#2656]
## ❔Question

![Screenshot from 2021-03-30 16-39-17](https://user-images.githubusercontent.com/20749514/112959950-86488480-9176-11eb-8332-270f2f2139d4.png)

## Additional context
Multi-GPU Memory-Usages are not equal. It is higher for the first one which becomes the bottleneck.

**Comment by glenn-jocher** [comment]
👋 Hello, thank you for your interest in our work! This issue seems to lack the minimum requirements for a proper response, or is insufficiently detailed for us to help you. Please note that most technical problems are due to:

- **Your modified or out-of-date code**. If your issue is not reproducible in a new `git clone` version of this repo we can not debug it. Before going further run this code and verify your issue persists:
```bash
$ git clone https://github.com/ultralytics/yolov5 yolov5_new  # clone latest
$ cd yolov5_new
$ python detect.py  # verify detection

# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible in one of our 3 common datasets ([COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [COCO128](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), or [VOC](https://github.com/ultralytics/yolov5/blob/master/data/voc.yaml)) we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of your labels and images.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, verify your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below. If in doubt, download Python 3.8.0 from https://www.python.org/, create a new [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), and install requirements.

If none of these apply to you, we suggest you close this issue and raise a new one using the 🐛 **Bug Report template**, providing screenshots and a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) of your issue. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.7`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab and Kaggle** notebooks with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>


## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by jeannotes** [comment]
@daikankan how do you solve this problem

**Comment by glenn-jocher** [comment]
@jeannotes @daikankan we see no serious imbalance between GPU memory utilization. This is an 8X A100 training on an AWS P4d instance with YOLOv5x6 as logged by W&B. Device 0 will always require slightly more memory as it maintains the model, EMA, etc.

<img width="858" alt="Screenshot 2021-06-07 at 17 29 43" src="https://user-images.githubusercontent.com/26833433/121045386-08c47480-c7b6-11eb-8ae6-d7d91e030585.png">

**Comment by glenn-jocher** [comment]
@jeannotes also regarding first epoch usage, we don't observe any abnormalities either. In our own public runs (8x A100 instances training multiple VOC models here for 50 epochs) you can see that there's no additional GPU memory used in the first epoch vs later epochs.
https://wandb.ai/glenn-jocher/VOC/runs/33a48d47/system?workspace=user-glenn-jocher
<img width="1559" alt="Screenshot 2021-04-22 at 17 21 46" src="https://user-images.githubusercontent.com/26833433/115740643-52d0d280-a38f-11eb-9b64-0a3928bc7658.png">

**Comment by wxupjack** [comment]
You can add `torch.cuda.empty_cache()` after each batch iteration.(train.py L359)
https://github.com/ultralytics/yolov5/blob/4effd064b169fc049b4a4bca401b120bf2e93c14/train.py#L355-L362
and L242 in val.py
https://github.com/ultralytics/yolov5/blob/4effd064b169fc049b4a4bca401b120bf2e93c14/val.py#L239-L245

However, I don't think this method is elegant. Although I can double the batch_size, it still slows a bit down training (not 2x faster). 

Also, at the end of val, the memory of master GPU become large again. For example, using the origin batch size, it takes 3GB. If I triple the size, it takes about 9GB in most of the train/val time, but, become over 12GB before val finished (then OOM error).

So I hope someone can help and give a better solution.

> envs:
> - yolov5 commit: e6e36aac109794999f1dafab244b9ec4887a33d1
> - torch:1.9.0
> - python:3.6.10
> - CUDA:11.1
> - Ubuntu:18.04

**Comment by glenn-jocher** [comment]
@wxupjack we see no serious imbalance between GPU memory utilization. This is an 8X A100 training on an AWS P4d instance with YOLOv5x6 as logged by W&B. Device 0 will always require slightly more memory as it maintains the model, EMA, etc.

<img width="858" alt="Screenshot 2021-06-07 at 17 29 43" src="https://user-images.githubusercontent.com/26833433/121045386-08c47480-c7b6-11eb-8ae6-d7d91e030585.png">

EDIT: For multi-GPU always train in our Docker image for best results with `torch.distributed.run` command.
